### PR TITLE
JS platform module override file

### PR DESCRIPTION
**Pull request by GantMan**
On the hope to integrate further with multiple operating systems, packager does the job!  

This solution can go slightly further and solve a growing concern, beyond OS specific JavaScript.  We have a growing need for OS VERSION specific javascript, in iOS and Windows.
### Case 1: Windows 7 (Platform added in this PR)

In the React Native Windows domain, we're creating a large shared JavaScript base, but in some situations (like Vibration), this requires alternate behavior.  This PR allows Windows 10 and Windows 7 to share most `*.windows.js` files, but allow for some specific `*.windows7.js` javascript files, by adding major versioning to the `getModule` cascade.
### Case 2: iOS 10

This could also help transition across difficult bridges of deprecation, as things like `RefreshControl` gets migrated to a first class citizen like it is in iOS10.  Different JS could load different native modules as beginning point for testing OS version specific solutions, or for breaking ground on other specific concerns for advanced operating systems, [discussed here](https://www.facebook.com/groups/reactnativeoss/permalink/1586734651623163/).  
### Summary:

This is a proposed solution, but the need already exists in ReactNativeWindows, which is currently developing support for Windows 7.

cc: @satya164  - Thoughts?

### CODE REVIEW

**Comment by bestander** [comment]
Ouch, I wish we would not need to hardcode extnesion of every platform.
We are having quite a lot of them now and this feature needs to become external.

@davidaurelio, what is your up-to-date position on more platforms support?
We are going to have more and more requests.

**Comment by GantMan** [comment]
@bestander - how about this?

The platform array looks for a possible local override file for additional platforms, and updated definition of the getModule functionality?   This would then allow other mediums like Apple TV etc to have their platform defined in the external lib, and leave FB with default configurations.  This keeps non-FB focussed platforms out of the management of FB projects.

File: packager/react-packager/src/node-haste/DependencyGraph/HasteMap.js [hunk 1]
```diff
@@ -96,12 +96,17 @@ class HasteMap extends EventEmitter {
...
 
-    // If platform is 'ios', we prefer .ios.js to .native.js which we prefer to
+    // If platform is 'ios10', we prefer .ios10.js to .ios.js
+    // which we prefer .ios.js to .native.js which we prefer to
     // a plain .js file.
     let module = undefined;
     if (module == null && platform != null) {
       module = modulesMap[platform];
     }
+    if (module == null && platform != null) {
```
Comment (bestander) [inline packager/react-packager/src/node-haste/DependencyGraph/HasteMap.js]: Why is this statement separate from the above?
Looks like the same condition

**Comment by bestander** [comment]
@GantMan, if everything can be with a local override then it looks like a perfect solution.
Can you try this out?
Better, if you see add some tests to prove how it works in more complex setups: ios + android + windows7 for example

File: packager/react-packager/src/node-haste/DependencyGraph/HasteMap.js (hunk 1 above)
Comment (GantMan) [inline packager/react-packager/src/node-haste/DependencyGraph/HasteMap.js]: it is the same condition, but the above line might have made module not null.  It's not an else if.  It tricked me up, too.   But this is the cascading code.   If the above code modified module, then the second code doesn't run.

I don't know who chose to write this code like this... feels unreadable, but I followed suite :)

"When in Rome"

**Comment by GantMan** [comment]
OK @bestander - Here's the updated PR with the platform agnostic config:

**Any update to platforms array is no-longer a PR to packager.**   

Each additional platform, like Windows, will implement a `.resolveConfig` in the root of their projects.  Here they can add platforms, and overrides needed.  This also place-marks a file for platform specific needs (like the hypothetical iOS10 discussion).

**This PR now checks for a config, and if none are found the default behavior is used.**

_e.g._ When Windows projects get created by `react-native windows` a `.resolveConfig` file will be placed in the project root.   Here additional platforms can be configured, and the HasteMap's `getModule` can be overridden.  Basically the non-invasive version of my previous needs.


## Appendix:

### User Experience
image

The file I created in the root of my project:
```javascript
module.exports = {
  cacheTransformResults: true,
  additionalPlatforms: ['windows7'],
  getModule: (hasteMap, name, preferNative, nativePlatform, genericPlatform, platform) => {
    const modulesMap = hasteMap._map[name];
    if (modulesMap == null) {
      return null;
    }

    // ['windows7.js', 'windows.js', 'native.js', '.js'];
    let module;
    if (module == null && platform != null) {
      module = modulesMap[platform];
    }
    if (module == null && platform != null) {
      const versionlessPlatform = platform.replace(/\d/, '');
      module = modulesMap[versionlessPlatform];
    }
    if (module == null && preferNative) {
      module = modulesMap[nativePlatform];
    }
    if (module == null) {
      module = modulesMap[genericPlatform];
    }
    return module;
  }
}
```

### Notes:  
* I left a log message in the Resolver for "Additional Platforms", which I think is nice.   It is a good litmus test if the config override file is present or not.  I figure as other platforms become more numerous, or existing platforms more independent, this log will be useful.
* I'll have the PR to Windows once this has been 👍

**Comment by GantMan** [comment]
additional note, I'm happy to update the initial PR title/description so that release notes align.  Just let me know when.

**Comment by GantMan** [comment]
@bestander  and @davidaurelio - please ignore the noise from merges.  I'm just making sure it doesn't get a conflict and stays up to date, considering I know this portion of the code changes.

**Comment by cpojer** [comment]
We should think more seriously about the ways we'd like to enable RNP to be configured. I'll bring it up with the team and get back to you here.

**Comment by GantMan** [comment]
@cpojer - just remember, perfect is the enemy of great.   Having an earmark for the beginning of configuration might be ok while we get this done.  I can tel you this is 100% needed in RN Windows.

**Comment by ide** [comment]
@GantMan One of the big things here is maintenance cost, it's often more costly to maintain code than it is to write it. If this isn't implemented in a way that @cpojer and his teammates feel they can be responsible for maintaining, you should consider using a fork for now (great) instead of landing upstream (perfect). I'm not saying this PR is or isn't maintainable, just explaining that merging code and taking on responsibility is a big concern.

**Comment by cpojer** [comment]
thanks for the explanation, @ide. @GantMan I definitely understand your concern but we need to make some more significant changes to packager to make it more scalable. I'd prefer to take things a bit more slowly from now on and find the right solutions that will be easy to support and maintain in the future, as @ide points out.

I think until we figure out some internal things it's best to use a fork.

**Comment by GantMan** [comment]
hey thanks for the feedback @cpojer  and @ide  👍  I completely understand it's best to keep things from getting hairy!!!  A lean packager is in everyone's interest.

That's kinda my goal. Please chat this out with me real quick.  You can't possibly keep merging into packager every time there's a new platform right?  Using a fork doesn't solve this, and it's a real problem that needs a solution.

**Here's the big question:  What's the best way to manage all these platforms and their needs?**

I'm not sure if I'm even close to the mark, but a config file that comes in from the platform (not from packager) makes sense to me.  This way you actually see the code living on a separate fork as desired.  Packager allows overrides for platforms, and releases you from further PRs.   It's convention over configuration, because with no config file, the defaults are used.  

I'm open to any other solution suggestions, but it's a problem that's going to keep bubbling up, and will need to get some kind of place to start.  So in theory, what's the best way?

**Comment by cpojer** [comment]
I haven't fully explored all possible solutions to this yet but I believe that adding a configuration entry to an application's package.json is the right way to go. Jest supports packager's way of resolving platforms and is customizable (see https://github.com/facebook/jest/blob/master/packages/jest-react-native/jest-preset.json#L2 ).

I'm thinking that we should start with a configuration option, like this:

```
// package.json
{
  "rnp": {
    "platforms": ["ios", "android", "windows"]
  }
}
```

What do you think (Note: I haven't run this by the rest of the team yet, although I do feel reasonably strongly about this)?

**Comment by GantMan** [comment]
looks good!  Should be trivial to adjust the direction of this PR on platforms. 

One issue I can see, is the use of json over a js file would limit the ability to add a function change.

One of the needs I have, that others may have is overriding needed functions.  For instance overriding `getModule` that is specific to my platform, but not to any other platform. 

Where would we keep these code overrides?  Would that be a file swap in `package.json`'s `rnp` config?  

I like the direction!  Happy to update the PR as needed once there's a good battle plan 👍

**Comment by ide** [comment]
@cpojer For RN the convention is to put configuration options in rn-cli.config.js, unless there's a good reason otherwise we should put new options there

**Comment by cpojer** [comment]
@GantMan for the future we are exploring how to hook in custom resolution algorithms so that RNP will be more easily extensible.

@ide is this an app specific file that we read? If we already have that, yea we should go with this.

**Comment by ide** [comment]
@cpojer yep, it's better than JSON too because you can write JS and the config fields are functions that return config values instead of static config values <3

**Comment by GantMan** [comment]
@ide and @cpojer - is it ok if I take a crack at this?  I can modify this PR to pertain to `rn-cli.config.js` instead.

**Comment by cpojer** [comment]
I think the list of platforms should be an array of strings for now, without any additional resolution code. We'll be throwing out some of this resolution code and rewrite it and I'd prefer not to have to maintain this at the moment. Does that work for you?

**Comment by GantMan** [comment]
@cpojer -  it doesn't, but no worries.  The cascade feature mentioned above, is the part that lead me to start contributing to packager.

**How's this sound?**  I'll actually be in the bay area next week visiting a client, and speaking with Microsoft.  Any interest in a meeting or a lunch?  I'd love to see if there's a future to me being able to contribute to your rewrite, and lend a helping hand in general.

cc: @ide

**Comment by cpojer** [comment]
Unfortunately I'm currently in London with the rest of the RNP team that is located here. :(

**Comment by GantMan** [comment]
No worries! maybe when I have Eric from Microsoft in the same room, we can do a virtual chat?  That way we can plan the best way moving forward?  Happy to assist if we find a way.

**Comment by andrewimm** [comment]
I've prepared a similar PR to support custom extensions for React VR (.vr.js), but this seems like a more thorough solution.
I'd agree that `rn-cli.config.js` is the proper place to configure this, since it's similar in nature to the other overrides that are already being placed there (asset extensions, etc).

@cpojer I know you're currently working with your team to plan changes for improving RNP, so please feel free to loop me in, since we're the latest consumer for this shared toolset.
This isn't a launch blocker for React VR, since we currently monkey-patch `defaults.js` with a postinstall script, but obviously that's not a great permanent solution 😂  looking forward to project-configurable extensions as more platforms emerge.

**Comment by cpojer** [comment]
Hey @GantMan,

are you still working on this? As said previously, I think we should start out with a simple "platforms" config option in the rn config file that is an Array of platforms.

**Comment by GantMan** [comment]
hey @cpojer !

My plan is to work with your plan 👍 

So, I know you're hoping to re-do the packager so it can work with platforms a bit more, and that's why this isn't good to merge.  My goal is to help you so I can still get advanced platform functionality, without it all having to go upstream.  What I'd love to do is meet and find out how I can be an alpha tester or contribute to the new packager so platforms can bolt-on needed functionality, like I'm hoping to add here.

Feel free to ping me in slack, I'll be available for any meeting.   Adding platforms to packager is good, but not the full goal of this PR.

**Comment by GantMan** [comment]
If you like I can start a second PR that adds that, if that's what works best.

Would still hope to talk strategy on next steps, though.

**Comment by cpojer** [comment]
We are going to keep doing GitHub first development of RNP and aren't doing anything in secret. The version of RNP inside of this repo is always the source of truth, so whatever you can do in this codebase to improve it, that is welcome. We should potentially split up what you wanted to do and the platform discussion into separate PRs and ship them individually.

**Comment by cpojer** [comment]
Nice, we just commented the same thing.

**Comment by andrewimm** [comment]
@GantMan @cpojer As mentioned earlier, I have a commit that adds this platforms array to rn-cli, which should be ready to go (I think). I'll have to re-test it, since it's been a couple weeks, but in the meantime I wouldn't mind creating a PR for it:
https://github.com/andrewimm/react-native/commit/30660421de44cddcfbb2c719ce8f228019971f10

**Comment by cpojer** [comment]
That is exactly it. Would you mind sending this internally at FB so we can immediately run all the internal tests on this as well? Maybe also add one or two tests?

**Comment by GantMan** [comment]
@andrewimm - looks like this would need platforms passed in every time you start packager.  Would you consider a local config like I somewhat started in this PR that would allow a user to declare a platform?  Or perhaps a good way to have a per project- set of defaults?

**Comment by andrewimm** [comment]
@gantman this includes 2 different paths: you can explicitly pass it in via the command line, or you can add it to your project's local `rn-cli.config.js`. For React VR, we're adding a '.vr.js' platform, and our `rn-cli` files include the following:

```js
...
getPlatforms() {
  return ['vr'];
}
...
```

**Comment by cpojer** [comment]
Why is it a problem to read platforms every time packager is started?

**Comment by GantMan** [comment]
@andrewimm - perfect!

**Comment by GantMan** [comment]
@andrewimm - Are you PRing that branch soon?

**Comment by andrewimm** [comment]
@GantMan I've submitted it through FB's internal code review tools, so that we can immediately check it against our internal tests and avoid the facebook-github-bot import. Shouldn't be too long now.

**Comment by GantMan** [comment]
@andrewimm - awesome, and thank you so much!  When that's in place I'll close this branch and re-evaluate how best to move forward.  Thanks again everybody!

**Comment by andrewimm** [comment]
Landed in 67828a52bdc15f5bf5300fd3eed8a5a4adbc9727

**Comment by GantMan** [comment]
@andrewimm - the `rn-cli.config.js` has to be in the parent dir of the CLI?

This will cause problems, as re-writing node modules will rewrite the cli which will lose changes provided by external platforms.  This should check the root of the project AND the cli with a merge.

**Pull request closed without merging**